### PR TITLE
Implement starknet_traceTransaction method

### LINKED ISSUE #312: starknet_traceTransaction
implement https://github.com/starkware-libs/starknet-specs/blob/v0.3.0/api/starknet_trace_api_openrpc.json#L11C22-L11C47

### PR DESCRIPTION
**Pull request by cicr99**
Closes #312

### CODE REVIEW

File: rpc/errors.go [hunk 1]
```diff
@@ -30,6 +32,10 @@ var (
 		code:    1,
 		message: "Failed to write transaction",
 	}
+	ErrNoTraceAvailable = &RPCError{
```
Comment (rianhughes) [inline rpc/errors.go:55]: The data field needs added to the errors: https://github.com/starkware-libs/starknet-specs/blob/v0.3.0/api/starknet_trace_api_openrpc.json#L330

File: rpc/trace_test.go [hunk 2]
```diff
@@ -11,31 +11,44 @@ import (
...
+		}
+		expectedrespRaw, err := os.ReadFile("./tests/0xff66e14fc6a96f3289203690f5f876cb4b608868e8549b5f6a90a21d4d6329.json")
+		require.NoError(t, err, "Error ReadFile for TestTraceTransaction")
+
+		err = json.Unmarshal(expectedrespRaw, &rawjson)
+		require.NoError(t, err, "Error unmarshalling testdata TestTraceTransaction")
+		expectedResp = rawjson.Result
+	}
+
 	type testSetType struct {
```
Comment (rianhughes) [inline rpc/trace_test.go:34]: Might be worth testing the errors

**Comment by rianhughes** [comment]
Also, the rpc tests are failing

File: rpc/trace.go [hunk 3]
```diff
@@ -2,16 +2,63 @@ package rpc
...
 	"context"
-	"errors"
+	"encoding/json"
 
 	"github.com/NethermindEth/juno/core/felt"
 )
 
-var ErrNotImplemented = errors.New("not implemented")
+// For a given executed transaction, return the trace of its execution, including internal calls
+func (provider *Provider) TransactionTrace(ctx context.Context, transactionHash *felt.Felt) (TxnTrace, error) {
```
Comment (rianhughes) [inline rpc/trace.go:11]: The return type should be *RPCError. 

When you return `noTraceAvailableError` in this function, what's actually returned is `noTraceAvailableError.Error()` (so we lose the code and data information).

File: rpc/trace_test.go [hunk 4]
```diff
@@ -11,30 +11,63 @@ import (
...
 	for _, test := range testSet {
-		var out string
-		err := do(context.Background(), testConfig.provider.c, test.MissingMethod, &out)
-
-		if err == nil || err.Error() != "Method Not Found" {
-			t.Fatalf("Method %s is now available, got %v\n", test.MissingMethod, err)
+		resp, err := testConfig.provider.TransactionTrace(context.Background(), test.TransactionHash)
+		if err != nil {
+			require.Equal(t, test.ExpectedError, err)
+		} else {
```
Comment (rianhughes) [inline rpc/trace_test.go:69]: Might be neater without the else

File: rpc/trace_test.go [hunk 5]
```diff
@@ -11,30 +11,63 @@ import (
...
 
 	for _, test := range testSet {
-		var out string
-		err := do(context.Background(), testConfig.provider.c, test.MissingMethod, &out)
-
-		if err == nil || err.Error() != "Method Not Found" {
-			t.Fatalf("Method %s is now available, got %v\n", test.MissingMethod, err)
+		resp, err := testConfig.provider.TransactionTrace(context.Background(), test.TransactionHash)
+		if err != nil {
+			require.Equal(t, test.ExpectedError, err)
```
Comment (rianhughes) [inline rpc/trace_test.go:68]: Note that you're comparing the `error` type, not `RPCError` (so the data and code fields aren't actually being checked)

File: rpc/errors.go [hunk 6]
```diff
@@ -2,6 +2,8 @@ package rpc
 
 import "errors"
 
+var ErrNotImplemented = errors.New("not implemented")
```
Comment (rianhughes) [inline rpc/errors.go:5]: I don't think this error is really needed anymore

File: rpc/trace_test.go (hunk 4 above)
Comment (cicr99) [reply rpc/trace_test.go:69]: This is funny! I pointed out the opposite over here https://github.com/NethermindEth/starknet.go/pull/332#discussion_r1327199608

File: rpc/trace.go (hunk 3 above)
Comment (cicr99) [reply rpc/trace.go:11]: I'm not sure that's right.
The mock_test method for this pr returns an `error` as well and in this [line](https://github.com/NethermindEth/starknet.go/pull/328/files#diff-617ac3083426917993255a7947979919a2700fce3d4d158586d30561ee91e396R697) I return an RPCError.  When I tried in the test to compare the resulting error and the expected one it also considered the other fields, besides the error message I mean. In fact it failed becuase the data field was different

File: rpc/trace_test.go (hunk 5 above)
Comment (cicr99) [reply rpc/trace_test.go:68]: I think this has to do with this same comment https://github.com/NethermindEth/starknet.go/pull/328#discussion_r1327217983. Let's settle that and then revisit this?

File: rpc/errors.go (hunk 6 above)
Comment (cicr99) [reply rpc/errors.go:5]: I thought so too when I removed it from the `trace.go` file, but turns out is also being used in `block_test.go`. The best place to put it was in the `errors.go` file

File: rpc/trace_test.go (hunk 4 above)
Comment (rianhughes) [reply rpc/trace_test.go:69]: Lol, I'm happy either way. Guess we all have our own styles :)

File: rpc/trace.go (hunk 3 above)
Comment (rianhughes) [reply rpc/trace.go:11]: You're right!! My bad

**Review by rianhughes — APPROVED** [review]

**Pull request merged**